### ISSUE TITLE
Python 3.11.0b4 - py launcher fails when running a script with a shebang line

### ISSUE THREAD

**Issue by pfmoore** [python/cpython#94772]
**Bug report**

When running a script with a "shebang" line, the `py.exe` launcher in Python 3.11.0b4 fails with an error:

To reproduce, create the following script as `x.py`:

```python
#!/usr/bin/env python
import sys
print(sys.version)
print(sys.executable)
```

Run the script using `py.exe` built from a 3.11.0b4 checkout using `build.bat`:

```
❯ .\amd64\py.exe .\x.py
 .\x.py': The system cannot find the file specified.ata\Local\Programs\Python\Python310\python.exe
```

Note the truncated python executable name. This appears to be related to the use of CRLF line endings in the file - if I switch to LF line endings, I get

```
❯ .\amd64\py.exe .\x.py
Unable to create process using 'C:\Users\Gustav\AppData\Local\Programs\Python\Python310\python.exe
 .\x.py': The system cannot find the file specified.
```

**Your environment**

- Launcher from 3.11.0b4, but the active Python version is 3.10.3.
- Operating system and architecture: Windows 11, 64-bit.

**Comment by pfmoore** [comment]
OK, I just did a bisect, and found the bad commit was bad86a621af61f383b9f06fe4a08f66245df99e2 "bpo-46566: Add new py.exe launcher implementation (GH-32062)". Sigh, I suppose that should have been rather obvious 🙂

@zooba Any insights as to what's wrong here? As it's a complete new implementation, I don't honestly know where to look...

@pablogsal This is a fairly serious issue (assuming it's reproducible elsewhere and isn't somehow tied to something peculiar on my system), so it may warrant being labelled as a release blocker. Sorry!

**Comment by pablogsal** [comment]
Added release blocker label for 3.11.0b5

**Comment by pfmoore** [comment]
I just found out that there's debug logging in the launcher. Debug info:

```
❯ $env:PYLAUNCHER_DEBUG=1  
❯ .\amd64\py.exe C:\Work\Projects\cpython\PCbuild\x.py
argv0: C:\Work\Projects\cpython\PCbuild\amd64\py.exe
version: 3.11.0b4
# Read 74 bytes from C:\Work\Projects\cpython\PCbuild\x.py to find shebang line
Found shebang: /Shebang: /usr/bin/env python

# Reading from C:\Users\Gustav\AppData\Local\py.ini for commands/python
# Did not find file C:\Users\Gustav\AppData\Local\py.ini
# Reading from C:\Work\Projects\cpython\PCbuild\amd64\py.ini for commands/python
# Did not find file C:\Work\Projects\cpython\PCbuild\amd64\py.ini
# Treating shebang command 'python' as 'py'
# Reading from C:\Users\Gustav\AppData\Local\py.ini for defaults/py_python
# Did not find file C:\Users\Gustav\AppData\Local\py.ini
# Reading from C:\Work\Projects\cpython\PCbuild\amd64\py.ini for defaults/py_python
# Did not find file C:\Work\Projects\cpython\PCbuild\amd64\py.ini
SearchInfo.originalCmdLine: "C:\Work\Projects\cpython\PCbuild\amd64\py.exe" C:\Work\Projects\cpython\PCbuild\x.py
SearchInfo.restOfCmdLine:  C:\Work\Projects\cpython\PCbuild\x.py
SearchInfo.executablePath: (null)
SearchInfo.scriptFile: C:\Work\Projects\cpython\PCbuild\x.py
SearchInfo.executable: python.exe
SearchInfo.executableArgs:

SearchInfo.company: (null)
SearchInfo.tag: (empty)
SearchInfo.oldStyleTag: True
SearchInfo.lowPriorityTag: False
SearchInfo.exclude32Bit: False
SearchInfo.only32Bit: False
SearchInfo.allowDefaults: True
SearchInfo.allowExecutableOverride: True
SearchInfo.windowed: False
SearchInfo.list: False
SearchInfo.listPaths: False
SearchInfo.help: False
 -V:3.10          C:\Users\Gustav\AppData\Local\Programs\Python\Python310\python.exe
 -V:3.9           C:\Users\Gustav\AppData\Local\Programs\Python\Python39\python.exe
 -V:3.8           C:\Users\Gustav\AppData\Local\Programs\Python\Python38\python.exe
env.company: PythonCore
env.tag: 3.10
# about to run: C:\Users\Gustav\AppData\Local\Programs\Python\Python310\python.exe
 C:\Work\Projects\cpython\PCbuild\x.py
Unable to create process using 'C:\Users\Gustav\AppData\Local\Programs\Python\Python310\python.exe
 C:\Work\Projects\cpython\PCbuild\x.py': The system cannot find the file specified.
```

**Comment by zooba** [comment]
I'm going to assume it's an off-by-one error in [this line](https://github.com/python/cpython/pull/32062/files#diff-6995823185ca496d08270284e34bb319a461c74bbf4bd6117df1438e2ea71492R861) or three lines down where we add 1 to calculate the length of the string.

Embarrassingly, the tests only test scripts with nothing after the shebang 🤦‍♂️ So adding a second line to the shebang tests will show the issue.

I probably won't get time to fix anything this week, so feel free, otherwise once I'm back at work I'll get to it.

**Comment by pfmoore** [comment]
lol, the three existing tests for the shebang all use a "script" that is a single line with no terminating newline. So they all fail with the fix I've made.

On the assumption that we want to support files with just a shebang and no newline or content (I'm not sure why we do, but I'm not going to make that call...) I'll update my fix *and* add some tests for the case where there is a newline...

**Comment by zooba** [comment]
I mean, we want them all to work, but I'd definitely rather have files with content also work. There shouldn't be any reason that both can't work though, it just makes the logic for stripping off trailing `\r\n` characters a bit more complex.

New tests would be great.

**Comment by pfmoore** [comment]
Done, and the PR has passed CI. If you have a chance to review, that would be great.

BTW, I found the logic in that bit of code pretty hard to follow. I didn't change it because I wanted to just do the minimal change necessary, but would you be OK with a follow-up to make the code a bit more verbose but (hopefully!) clearer?
